test/xbgpu/test_engine.py

```diff
@@ -866,32 +919,58 @@ async def _send_data(
                     f"actual: {ig_recv['frequency'].value}."
                 )
 
-                corrprod_results[i][j] = ig_recv["xeng_raw"].value
+                corrprod_results[corrprod_output.name][j] = ig_recv["xeng_raw"].value
+
+            # Confirm that there are no more accumulations to receive
+            with pytest.raises(spead2.Stopped):
+                await self.get_data_heap(stream, ig_recv)
+
+        beam_results: dict[str, np.ndarray] = {}
+        if beam_capture_stop_heap_indices is None:
+            # Rather be explicit for each `beam_output` to simplify the logic below
+            beam_capture_stop_heap_indices = [None] * len(beam_outputs)
+        # We adjust the corresponding beam stream to only receive the required
+        # amount of data (up until the capture-stop point).
+        for beam_output, capture_stop_heap_index in zip(beam_outputs, beam_capture_stop_heap_indices):
+            if capture_stop_heap_index is not None:
+                # Obtaining a subset using zero (0) would result in an empty array.
+                # We also don't want to use negative indices to work from the end
+                # of the list of indices. Positive values are desired and more
+                # explicit in this case.
+                beam_results[beam_output.name] = np.zeros(
+                    (
+                        capture_stop_heap_index,
+                        n_channels_per_substream,
+                        n_spectra_per_heap,
+                        COMPLEX,
+                    ),
+                    bsend.SEND_DTYPE,
+                )
+            else:
+                beam_results[beam_output.name] = np.zeros(
+                    (len(batch_indices), n_channels_per_substream, n_spectra_per_heap, COMPLEX),
+                    bsend.SEND_DTYPE,
+                )
```

Comment:
This code is mostly a copy of the `if` branch. I'd suggest you instead do something like
```python
n_batches = len(batch_indices) if capture_stop_heap_index is None else capture_stop_heap_index
beam_results[beam_output.name] = np.zeros(
    (n_batches, n_channels_per_substream, n_spectra_per_heap, COMPLEX),
    bsend.SEND_DTYPE,
)
```
(untested)